### PR TITLE
Add `modify_fields` processor and use it to set `logName`

### PR DESCRIPTION
**Pull request by quentinmit**
Major changes in this branch:

* Refactor `exclude_logs` and filter parsing to be more consistent with Cloud Logging syntax
* Use Lua for most `exclude_logs` and `modify_fields` conditions
* Add `modify_fields` processor
  * Source value: `move_from`, `copy_from`, `static_value`
  * `default_value` if source value is unset
  * `map_values`, `type` to transform the values
  * `omit_if` to conditionally remove a field
* Set `logName` using the new processor

### CODE REVIEW

File: confgenerator/filter/filter.go
Comment (jefferbrecht) [inline confgenerator/filter/filter.go:67]: Comment and function name mismatch

File: confgenerator/confgenerator_test.go
Comment (jefferbrecht) [reply confgenerator/confgenerator_test.go:46]: They're sort of bundled since `modify_fields` adds golden Lua files to the tests. Deferring the refactor would mean either not including those tests here or doing a different/temporary refactor to accommodate the Lua files.

Comment (qingling128) [reply confgenerator/confgenerator_test.go:46]: Oh I meant to split it into a separate PR and get that one in first.

File: confgenerator/logging_modify_fields.go
Comment (quentinmit) [reply confgenerator/logging_modify_fields.go:54]: These `panic` errors should never be thrown, ever. They indicate a bug in Ops Agent that we failed to catch through unit tests. All of this code only runs at config generation time, so yes, it can't ever panic due to the contents of a log entry.

There won't be any logging if fields are missing; it treats "missing" as a valid value, so trying to copy a missing field will just unset the new field, condition matching a missing field will evaluate false, etc. I think that's reasonable logic.

File: confgenerator/confgenerator_test.go
Comment (quentinmit) [reply confgenerator/confgenerator_test.go:224]: Uhh, there is no change in actual behavior here. The unit tests already have a bunch of checked-in files named `golden_$foo` that represent the expected content of a file named `$foo`. This comment is saying that error messages will be treated as a file called `error`, which means the expected value will be in `golden_error` as it already was. The files aren't written to disk; they're represented by an in-memory map of files and contents. I think it's still reasonable to call it a file.

File: confgenerator/confgenerator_test.go
Comment (quentinmit) [reply confgenerator/confgenerator_test.go:182]: I don't think so, the calling code will log this as a test failure.

File: confgenerator/confgenerator_test.go
Comment (quentinmit) [reply confgenerator/confgenerator_test.go:193]: Yeah, this fits into the `got` / `want` test pattern. The map represents the `want` state - i.e. the actual expected filenames. The `got` state is the map returned by `GenerateFluentBitConfigs`.

File: confgenerator/confgenerator_test.go
Comment (quentinmit) [reply confgenerator/confgenerator_test.go:127]: `MergeConfFiles` may have failed to generate the config files and returned an error, or it might have generated one or both of the config files and then returned an error. We don't know if it did generate any files if it returns an error.

The code does not leave the two files around if the test fails. It always removes them.

File: confgenerator/confgenerator_test.go
Comment (quentinmit) [reply confgenerator/confgenerator_test.go:165]: It means that there's a `golden_foo.conf` file on disk but the test did not generate `foo.conf`. This now happens frequently due to the Lua filters being generated.

File: confgenerator/logging_modify_fields.go
Comment (quentinmit) [reply confgenerator/logging_modify_fields.go:113]: It means that this needs to be implemented before this PR is ready to merge :)

File: confgenerator/logging_modify_fields.go
Comment (quentinmit) [reply confgenerator/logging_modify_fields.go:30]: `string`'s default value is `""`. `*string`'s default value is `nil`. You need the latter to be able to distinguish between an empty config and `static_value: ""`; this lets the user set an empty string as the value of a label.

File: confgenerator/logging_modify_fields.go
Comment (quentinmit) [reply confgenerator/logging_modify_fields.go:28]: Nope, because they're not required. None of the fields being specified is treated as `copy_from` the same metric; i.e. it reads the value from this field, applies the transformations, and writes back to the same field.

File: confgenerator/logging_modify_fields.go
Comment (quentinmit) [reply confgenerator/logging_modify_fields.go:33]: This is an unexported field and can't be set via YAML. This is just the Lua variable name that contains the field's value and is set and used by the code below.

File: integration_test/ops_agent_test.go
Comment (quentinmit) [reply integration_test/ops_agent_test.go:594]: I will add those when `omit_if` is implemented. :)

File: integration_test/ops_agent_test.go
Comment (quentinmit) [reply integration_test/ops_agent_test.go:630]: I'm holding off on that until we've decided exactly what shape the field should take...

File: confgenerator/testdata/invalid/linux/logging-processor_exclude_logs_invalid_filter/golden_error
Comment (quentinmit) [reply confgenerator/testdata/invalid/linux/logging-processor_exclude_logs_invalid_filter/golden_error:1]: This is an exclude_logs test that you're commenting on, so the other validation doesn't apply :) I tweaked the text a bit while still following the error message style.

File: confgenerator/logging_modify_fields.go
Comment (quentinmit) [reply confgenerator/logging_modify_fields.go:27]: All of the bullets you list are already being validated; are you asking for more invalid test cases to cover that validation?

File: confgenerator/filter/internal/ast/ast.go
Comment (quentinmit) [reply confgenerator/filter/internal/ast/ast.go:242]: I wasn't sure why it was like this so I wanted to keep it untouched for this PR. Hopefully we can stop doing this if they expose regex to Lua...

File: confgenerator/filter/internal/ast/ast.go
Comment (quentinmit) [reply confgenerator/filter/internal/ast/ast.go]: Sure, `panic` calls that can leave a package are only allowed when they indicate a clear programming error. I think this qualifies (the validation/parsing code shouldn't allow any other value to end up in the `Operator` field).

File: confgenerator/confgenerator.go
Comment (quentinmit) [reply confgenerator/confgenerator.go]: Added a comment. I don't think this calls for a named return variable (it can get confusing when a function has a named `error` result and also contains statements like `foo, err := funcCall()` because that might create a new `err` that shadows the return variable).

File: confgenerator/logging_modify_fields.go
Comment (qingling128) [reply confgenerator/logging_modify_fields.go:27]: Yes, I was thinking about some unit tests for the invalid cases.

File: confgenerator/logging_modify_fields.go
Comment (qingling128) [reply confgenerator/logging_modify_fields.go:28]: To make sure I understand it fully, what is an example config that does not have any of `MoveFrom`, `CopyFrom` or `StaticValue`?

File: confgenerator/logging_modify_fields.go
Comment (qingling128) [inline confgenerator/logging_modify_fields.go:30]:
```suggestion
        // Use "*string" instead of "string" here so the default is nil instead of an empty config.
        // This is required to allow users to specify an empty string as the config value.
	StaticValue *string `yaml:"static_value" validate:"excluded_with=MoveFrom CopyFrom"`
```

File: confgenerator/logging_modify_fields.go
Comment (qingling128) [reply confgenerator/logging_modify_fields.go:30]: Ah I see. Thanks for the clarification. Sent a suggested comment for future readers.

File: confgenerator/logging_modify_fields.go
Comment (qingling128) [reply confgenerator/logging_modify_fields.go:54]: Good to know.

File: confgenerator/logging_modify_fields.go
Comment (qingling128) [reply confgenerator/logging_modify_fields.go:113]: Ah I see. I'll leave this one open for now.

File: confgenerator/confgenerator_test.go
Comment (qingling128) [reply confgenerator/confgenerator_test.go:128]: Sent a suggested edit in https://github.com/GoogleCloudPlatform/ops-agent/pull/514#discussion_r843149562. Resolving this thread.

File: confgenerator/confgenerator_test.go
Comment (qingling128) [reply confgenerator/confgenerator_test.go:166]: Sent a suggested edit: https://github.com/GoogleCloudPlatform/ops-agent/pull/514#discussion_r843176868.

Resolving this one.

File: confgenerator/testdata/invalid/linux/logging-processor_exclude_logs_invalid_path_cr/golden_error
Comment (jefferbrecht) [inline confgenerator/testdata/invalid/linux/logging-processor_exclude_logs_invalid_path_cr/golden_error:1]: I can't immediately tell which change caused this, but is there any chance we can get the escaping back in these error messages? Not a blocker but the `\r` is much easier to see when it's escaped.

File: confgenerator/testdata/invalid/linux/logging-processor_exclude_logs_invalid_source/input.yaml
Comment (jefferbrecht) [inline confgenerator/testdata/invalid/linux/logging-processor_exclude_logs_invalid_source/input.yaml:11]: s|`confgenerator/testdata/invalid/linux/logging-processor_exclude_logs_invalid_source`|`confgenerator/testdata/invalid/linux/logging-processor_modify_fields_invalid_source`

File: confgenerator/testdata/invalid/linux/logging-processor_exclude_logs_multiple_sources/input.yaml
Comment (jefferbrecht) [inline confgenerator/testdata/invalid/linux/logging-processor_exclude_logs_multiple_sources/input.yaml:12]: `mv confgenerator/testdata/invalid/linux/logging-processor_exclude_logs_multiple_sources/* confgenerator/testdata/invalid/linux/logging-processor_modify_fields_multiple_sources`

File: confgenerator/testdata/valid/linux/logging-processor_exclude_logs/input.yaml
Comment (jefferbrecht) [inline confgenerator/testdata/valid/linux/logging-processor_exclude_logs/input.yaml:21]: Why were these changed? No issue, just curious.

File: confgenerator/testdata/invalid/linux/logging-processor_exclude_logs_invalid_path_cr/golden_error
Comment (quentinmit) [reply confgenerator/testdata/invalid/linux/logging-processor_exclude_logs_invalid_path_cr/golden_error:1]: I agree, it was the change to make `fluentBitPath` return an unescaped path.

File: confgenerator/logging_modify_fields.go
Comment (quentinmit) [reply confgenerator/logging_modify_fields.go]: Hm, I don't want to put this here and interrupt the flow of the code for this note. Using `*string` in this way is a common convention - e.g. every protobuf does this. Can I get away with leaving out a comment for this?

File: confgenerator/logging_modify_fields.go
Comment (quentinmit) [reply confgenerator/logging_modify_fields.go:27]: I added more unit tests now.

File: confgenerator/filter/internal/ast/ast.go
Comment (jefferbrecht) [inline confgenerator/filter/internal/ast/ast.go:31]:
```suggestion
// text may contain a small number of special characters, must start
```

File: confgenerator/testdata/valid/linux/all-backward_compatible_with_explicit_exporters/golden_e59640aecc0edba3f991865e880afb46.lua
Comment (qingling128) [reply confgenerator/testdata/valid/linux/all-backward_compatible_with_explicit_exporters/golden_e59640aecc0edba3f991865e880afb46.lua]: Sync'd offline. For cases where a LUA script is shared across multiple pipelines, prefixing the `md5sum` with `{pipeline_id}.{RECEIVER_ID}` would cause duplication and complexity.

Given that tradeoff, we decided to just bear with it for the debugging / troubleshooting inconvenience.

Hopefully this would become moot when https://github.com/fluent/fluent-bit/pull/5302 lands.

File: integration_test/ops_agent_test.go
Comment (qingling128) [reply integration_test/ops_agent_test.go:623]: This has been reflected in the `default_value` implementation and related tests.

File: confgenerator/logging_modify_fields.go
Comment (qingling128) [reply confgenerator/logging_modify_fields.go]: Got it.

**Review by qingling128 — APPROVED** [review]
Did another pass and resolved some open threads. The only remaining things:
* Two minor suggested edits https://github.com/GoogleCloudPlatform/ops-agent/pull/474#pullrequestreview-941033924[](https://github.com/GoogleCloudPlatform/ops-agent/pull/474#pullrequestreview-941033924)
* `instrumentation_source` related naming conversation: We can split `instrumentation_source` to a separate PR to get this one unblocked first.

LGTM modulo the two things above are addressed.

**Review by qingling128 — APPROVED** [review]
LGTM

**Pull request merged**
